pkg/controller/garbagecollector/memcachediscovery/client.go

```diff
@@ -0,0 +1,185 @@
+		d.lock.RLock()
+		defer d.lock.RUnlock()
+		cachedVal, ok := d.groupToServerResources[groupVersion]
+		return cachedVal, ok
+	}(); ok {
+		return rl, nil
+	}
+
+	liveResources, err := d.delegate.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		glog.V(3).Infof("skipped caching discovery info due to %v", err)
+		return nil, err
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.groupToServerResources[groupVersion] = liveResources
+	return liveResources, nil
+}
+
+// ServerResources returns the supported resources for all groups and versions.
+func (d *Client) ServerResources() ([]*metav1.APIResourceList, error) {
+	apiGroups, err := d.ServerGroups()
+	if err != nil {
+		return nil, err
+	}
+	groupVersions := metav1.ExtractGroupVersions(apiGroups)
+	result := []*metav1.APIResourceList{}
+	for _, groupVersion := range groupVersions {
+		resources, err := d.ServerResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, resources)
+	}
+	return result, nil
+}
+
+func (d *Client) ServerGroups() (*metav1.APIGroupList, error) {
+	if g, ok := func() (*metav1.APIGroupList, bool) {
+		d.lock.RLock()
+		defer d.lock.RUnlock()
+		return d.groupList, d.groupList != nil
+	}(); ok {
+		return g, nil
+	}
+
+	liveGroups, err := d.delegate.ServerGroups()
+	if err != nil {
+		glog.V(3).Infof("skipped caching discovery info due to %v", err)
+		return nil, err
+	}
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.groupList = liveGroups
+	return liveGroups, nil
+}
+
+func (d *Client) RESTClient() restclient.Interface {
+	return d.delegate.RESTClient()
+}
+
+func (d *Client) ServerPreferredResources() ([]*metav1.APIResourceList, error) {
+	return d.delegate.ServerPreferredResources()
+}
+
+func (d *Client) ServerPreferredNamespacedResources() ([]*metav1.APIResourceList, error) {
+	return d.delegate.ServerPreferredNamespacedResources()
+}
+
+func (d *Client) ServerVersion() (*version.Info, error) {
+	return d.delegate.ServerVersion()
+}
+
+func (d *Client) SwaggerSchema(version schema.GroupVersion) (*swagger.ApiDeclaration, error) {
+	return d.delegate.SwaggerSchema(version)
+}
+
+func (d *Client) OpenAPISchema() (*spec.Swagger, error) {
+	return d.delegate.OpenAPISchema()
+}
+
+func (d *Client) Fresh() bool {
+	return true
+}
+
+// Invalidate refreshes the cache, blocking calls until the cache
+func (d *Client) Invalidate() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.groupToServerResources, d.groupList = d.refresh()
+}
+
+// Refresh reloads the cached discovery information, continuing to serve the
+// previously cached values until it has finished getting the new values.
+func (d *Client) Refresh() {
+	rl, gl := func() (map[string]*metav1.APIResourceList, *metav1.APIGroupList) {
+		d.lock.RLock()
+		defer d.lock.RUnlock()
+		return d.refresh()
+	}()
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.groupToServerResources, d.groupList = rl, gl
+}
+
+// d must be RLocked or Locked.
```

Comment:
why?  Just for the cached return on error?  I'd rather see that handled by outter logic that handles the error.